Add root and no_recommends parameters in the public interface for Zypper and RPM (lowpkg)

aplanas opened the pull request:
### What does this PR do?
Provides a new root parameter in the public interfaces for rpm and zypper

### What issues does this PR fix or reference?
New feature

### Tests written?

For the low level RPM yes, for the zypper level I adjusted the current tests.

aplanas committed:
zypper: add root configuration parameter

aplanas committed:
Fix typo in comment

aplanas committed:
lowpkg: add parameter to change root directory

The CLI rpm command allows the --root parameter to change the
expected location where the rpm database can be found.

This patch add a new optional parameter in the public interface
to allow the set of the new root location.

Update the tests to use the extra parameter.

aplanas committed:
Add root parameter into the zypper module

The zypper CLI provides a way to change the path where zypper expect
to find the required configuration files and repositories.

This feature is useful to bootstrap chroot environments, inspect
repositories and packages from locally mounted devices, or help
during the installation of a new OS from the SUSE family.

This patch add the root optional parameter for each command in the
public interface, and fix the tests.

isbm commented on salt/modules/rpm.py:
```diff
@@ -338,10 +357,13 @@ def file_list(*packages):
+        cmd.append('-qla')
     else:
-        cmd = ['rpm', '-ql']
+        cmd.append('-ql')
```
A bit spaggettish, maybe we can always add `-ql` and only append another `-a` when no packages? Or so:

```python
cmd.append('-ql{}'.format(packages and 'a' or ''))
```

Or like you did later:

```python
cmd.append('-ql' if packages else '-qla')
```

isbm commented on salt/modules/zypper.py:
```diff
@@ -2113,18 +2222,20 @@ def list_products(all=False, refresh=False):
-    OEM_PATH = "/var/lib/suseRegister/OEM"
+    oem_path = '/var/lib/suseRegister/OEM'
+    if root:
+        oem_path = os.path.join(root, oem_path)
```
Why constant became variable? This should be uppercase.

isbm commented on salt/modules/zypper.py:
```diff
@@ -2206,22 +2322,27 @@ def download(*packages, **kwargs):
-    CACHE_DIR = '/var/cache/zypp/packages/'
+    cache_dir = '/var/cache/zypp/packages/'
+    if root:
+        cache_dir = os.path.join(root, cache_dir)
```
Constant became variable. Please keep it upper-case.

isbm commented on salt/modules/zypper.py:
```diff
@@ -2359,7 +2494,7 @@ def list_provides(**kwargs):
 
 
-def resolve_capabilities(pkgs, refresh, **kwargs):
+def resolve_capabilities(pkgs, refresh=False, root=False, **kwargs):
```
Why sometimes `root` is `None` and sometimes `False`? This should be consistent.

isbm replied:
@meaksh is that OK to default `refresh` here?

aplanas committed:
pkg: Transfer optional parameters to lower levels.

aplanas committed:
pkgrepo: Transfer optional parameters to lower levels.

isbm commented on salt/modules/zypper.py:
```diff
@@ -2206,22 +2322,27 @@ def download(*packages, **kwargs):
 
     ret = {}
-    for root, dirnames, filenames in salt.utils.path.os_walk(CACHE_DIR):
+    for root, dirnames, filenames in salt.utils.path.os_walk(cache_dir):
```
Constant still variable...

aplanas replied:
Looks like that github is a bit inconsistent today, but as commented I am changing CACHE_DIR, so is not a constant anymore. I would find confusing something like this:

```
def my_func():
  A = 10
  if cond():
    A = A + 5
```

I change A, but the reader do not expect that I do that based on the uppercase. Changing the the case will fix the expectation.

aplanas committed:
zypper: fix the reset after the call

_Zypper class take note when a .call() is done, to clean up the data
when we access to some attribute.

This produces a bug when two calls are one after another an we set
some attributes via the __call__ method, as whatever is set will be
cleared after the first attribute is accessed.

For example:

zypper.attrib.call(..)
zypper(root=root).otherattrib.call(..)

The first call will set __called as True, and the reset of the inner
state of zypper will be cleared when otherattrib is accessed,
cleanning the status for __root.

This patch makes sure to clean the status also during the __call__
method, avoiding the cleanning when the attribute is accessed.

aplanas committed:
zypper: add no_recommends parameter

Add no_recommends parameter to install and upgrade actions.

aplanas commented:
@isbm ready for a deeper review!

aplanas committed:
zypper: fix pylint complain

rallytime commented:
@aplanas There are some tests failing with this change. Can you take a look?

https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py2/job/PR-50125/12/

aplanas commented:
@rallytime indeed, thanks

isbm commented on salt/modules/rpm.py:
```diff
@@ -128,21 +128,27 @@ def bin_pkg_info(path, saltenv='base'):
-           '--queryformat', r'%{NAME} %{VERSION}\n']
+    cmd = ['rpm']
+    if 'root' in kwargs:
+        cmd.extend(['--root', kwargs['root']])
```
This is bug. Correctly would be:
```python
if kwargs.get('root'):
   ...
```

aplanas replied:
Oh, you are referring to the line up. I will change. But still I do not see the error. Is there a case where __getitem__ is not defined for kwargs?

aplanas replied:
Ah nono. I get it now, the goal is to make sure that is not None , '' or 0. But this pattern can exclude 0, and sometimes is a valid value.

isbm commented on salt/modules/rpm.py:
```diff
@@ -199,6 +208,8 @@ def verify(*packages, **kwargs):
             verify_options = [x.strip() for x in six.text_type(verify_options).split(',')]
 
     cmd = ['rpm']
+    if 'root' in kwargs:
```
Same bug.

isbm commented on salt/modules/rpm.py:
```diff
@@ -266,10 +280,12 @@ def modified(*packages, **flags):
-        output_loglevel='trace',
-        python_shell=False)
+    cmd = ['rpm']
+    if 'root' in flags:
```
Bug. Check for root _value_.

isbm commented on salt/modules/rpm.py:
```diff
@@ -338,25 +357,31 @@ def file_list(*packages):
-    else:
-        cmd = ['rpm', '-ql']
+    cmd = ['rpm']
+    if 'root' in kwargs:
```
Here too.

isbm reviewed (changes requested):
@aplanas OK, I stopped, but there are more. Please first fix these. :wink:

aplanas commented:
@isbm I am done with the 'in's. Thanks!

isbm commented on salt/modules/zypper.py:
```diff
@@ -906,23 +950,25 @@ def _is_match(pkgname):
     repos_cfg = configparser.ConfigParser()
-    repos_cfg.read([REPOS + '/' + fname for fname in os.listdir(REPOS) if fname.endswith(".repo")])
+    if os.path.exists(repos):
+        repos_cfg.read([repos + '/' + fname for fname in os.listdir(repos) if fname.endswith(".repo")])
```
Add error logging here if `repos != REPOS` and not found.

isbm commented on salt/modules/zypper.py:
```diff
@@ -1029,7 +1087,8 @@ def mod_repo(repo, **kwargs):
 
-    repos_cfg = _get_configured_repos()
+    root = kwargs.get('root', None)
+    repos_cfg = _get_configured_repos(root=root)
```
This is a bug. 
```python
repos_cfg = _get_configured_repos(root=kwargs.get('root') or None)
```

aplanas replied:
I need to assign to root, as is used later. Also I am not able to see the bug, as root will be None in case that is not there.

I changed the code to some proposal that was before, also I am not able to see the bug.

isbm commented on salt/modules/zypper.py:
```diff
@@ -1702,8 +1796,9 @@ def list_locks():
-        with salt.utils.files.fopen(LOCKS) as fhr:
+    _locks = os.path.join(root, os.path.relpath(LOCKS, start='/')) if root else LOCKS
+    if os.path.exists(_locks):
+        with salt.utils.files.fopen(_locks) as fhr:
```
I'd rather try/except it instead of stat file twice.

aplanas replied:
done, complicate the code a bit

isbm commented on salt/modules/zypper.py:
```diff
@@ -1729,10 +1827,11 @@ def clean_locks():
     LCK = "removed"
     out = {LCK: 0}
-    if not os.path.exists("/etc/zypp/locks"):
+    locks = os.path.join(root, os.path.relpath(LOCKS, start='/')) if root else LOCKS
```
Technically, `os.path.sep`, not `'/'`. :wink:

aplanas replied:
indeed!

aplanas commented on salt/modules/zypper.py:
```diff
@@ -127,18 +127,28 @@ def _reset(self):
             self.__ignore_repo_failure = kwargs['no_repo_failure']
         if 'systemd_scope' in kwargs:
             self.__systemd_scope = kwargs['systemd_scope']
+        if 'root' in kwargs:
```
On a second though, are you sure? This check if to assign self.__root later in the same fashion that 'no_repo_failure' and 'systemd_scope'

isbm replied:
Well, OK, copypaste. Maybe let's to as is. But I still think the other parts also needs to be type-safe.

aplanas commented:
@isbm any more reviews? Any thing that I can do?

isbm reviewed (approved):
@aplanas nope, good to go. :wink:

aplanas commented on salt/modules/zypper.py:
```diff
@@ -1186,6 +1250,9 @@ def install(name=None,
             version=None,
             ignore_repo_failure=False,
+            no_recommends=False,
+            auto_agree_with_licenses=False,
```
Ah yes. Comes from a different place. Fixed.

aplanas commented:
This PR is reaching the 1 month old mark. If there anything that I can do the help here?

The code do not change the default behavior, as I did not change the semantic of the tests.

rallytime commented:
@terminalmage and @meaksh How does this look to you now?

aplanas commented:
thanks a lot for the reviews!
